### ISSUE TITLE
Network policy and hostNetwork pods do not work together

### ISSUE THREAD

**Issue by fxposter** [projectcalico/felix#1361]
**TLDR: network policies do not apply to the pods configured with `hostNetwork: true`**

### Steps to reproduce

Creating a test namespace and defaulting it to `deny` all traffic:
```
$ kubectl create ns policy-demo
$ kubectl annotate ns policy-demo "net.beta.kubernetes.io/network-policy={\"ingress\":{\"isolation\":\"DefaultDeny\"}}"
```

**SUCCESS: non-hostNetwork to non-hostNetwork**

Creating and exposing non-hostNetwork nginx deployment:
```
$ kubectl apply -f - <<EOF
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: nginx
  namespace: policy-demo
spec:
  replicas: 1
  template:
    metadata:
      labels:
        run: nginx
    spec:
      containers:
      - name: nginx
        image: nginx
        ports:
        - containerPort: 80
EOF

$ kubectl expose --namespace=policy-demo deployment nginx --port=80
```

Trying to access nginx from another non-hostNetwork pod:
```
$ kubectl run --namespace=policy-demo access -i --rm --image busybox --restart Never -- wget -T 5 nginx -O -
Connecting to nginx (10.100.0.35:80)
wget: download timed out
```

Everything works fine.

**FAILURE: non-hostNetwork to hostNetwork**

Adding `hostNetwork: true` to nginx:
```
$ kubectl apply -f - <<EOF
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: nginx
  namespace: policy-demo
spec:
  replicas: 1
  template:
    metadata:
      labels:
        run: nginx
    spec:
      hostNetwork: true
      containers:
      - name: nginx
        image: nginx
        ports:
        - containerPort: 80
EOF
```

Trying to reach nginx from non-hostNetwork pod:
```
$ kubectl run --namespace=policy-demo access -i --rm --image busybox --restart Never -- wget -q -T 5 nginx -O - | grep title
<title>Welcome to nginx!</title>
```

Works, but should not.

**FAILURE: hostNetwork to non-hostNetwork**

Nginx to `hostNetwork: false`:
```
$ kubectl apply -f - <<EOF
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: access
  namespace: policy-demo
spec:
  replicas: 1
  template:
    metadata:
      labels:
        run: access
    spec:
      hostNetwork: true
      dnsPolicy: ClusterFirst
      containers:
      - name: busybox
        image: busybox
        command: ["sleep", "10000"]
EOF
```

As it's not possible to `kubectl run` a pod with `hostNetwork: true` - creating a deployment with `hostNetwork: true`, which does nothing and just sleeps:
```
$ kubectl apply -f - <<EOF
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: nginx
  namespace: policy-demo
spec:
  replicas: 1
  template:
    metadata:
      labels:
        run: nginx
    spec:
      containers:
      - name: nginx
        image: nginx
        ports:
        - containerPort: 80
EOF
```

Getting the "access" pod name (for later `kubectl exec`):
```
$ ACCESS_POD_NAME=`kubectl get pods -n policy-demo -o 'go-template={{(index .items 0).metadata.name}}' -l 'run=access'`
```

Getting nginx ips (via the service and directly on host), because we no longer can `wget nginx`, because hostNetwork pods' do not have proper /etc/resolv.conf and do not contact the cluster DNS:
```
$ NGINX_SERVICE_IP=`kubectl get services -n policy-demo -o 'go-template={{(index .items 0).spec.clusterIP}}'`
$ NGINX_POD_IP=`kubectl get pods -n policy-demo -o 'go-template={{(index .items 0).status.podIP}}' -l 'run=nginx'`
```

Trying to reach nginx from pod via both IPs:
```
$ kubectl exec $ACCESS_POD_NAME -n policy-demo -- wget -q -T 5 $NGINX_SERVICE_IP -O - | grep title
<title>Welcome to nginx!</title>
$ kubectl exec $ACCESS_POD_NAME -n policy-demo -- wget -q -T 5 $NGINX_POD_IP -O - | grep title
<title>Welcome to nginx!</title>
```

Works, but should not.

**FAILURE: hostNetwork to hostNetwork**

Nginx to `hostNetwork: true`:
```
$ kubectl apply -f - <<EOF
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: nginx
  namespace: policy-demo
spec:
  replicas: 1
  template:
    metadata:
      labels:
        run: nginx
    spec:
      hostNetwork: true
      containers:
      - name: nginx
        image: nginx
        ports:
        - containerPort: 80
EOF
```

Trying to reach nginx again:
```
$ NGINX_SERVICE_IP=`kubectl get services -n policy-demo -o 'go-template={{(index .items 0).spec.clusterIP}}'`
$ NGINX_POD_IP=`kubectl get pods -n policy-demo -o 'go-template={{(index .items 0).status.podIP}}' -l 'run=nginx'`

$ kubectl exec $ACCESS_POD_NAME -n policy-demo -- wget -q -T 5 $NGINX_SERVICE_IP -O - | grep title
<title>Welcome to nginx!</title>
$ kubectl exec $ACCESS_POD_NAME -n policy-demo -- wget -q -T 5 $NGINX_POD_IP -O - | grep title
<title>Welcome to nginx!</title>
```

Also works, while should not.

-----

All those results are observed in Vagrant (from http://docs.projectcalico.org/v2.0/getting-started/kubernetes/installation/vagrant/).

Are those the expected results (ie: Calico is not going to support such cases automatically) or bugs?

**Comment by fasaxc** [comment]
This is not expected to work.  Setting `hostNetwork: true` is saying "don't isolate this pod, treat it as a trusted part of the host".

However, Calico does support applying policy to the host's `eth0`, for example, by using `hostEndpoints`.

**Comment by fxposter** [comment]
Thanks!
